In [1]:
import pandas as pd

Problem Statement: https://www.kaggle.com/c/house-prices-advanced-regression-techniques
Dataset: https://github.com/abulbasar/data/tree/master/kaggle-houseprice

In [2]:
df = pd.read_csv("/data/kaggle/data_combined_cleaned.csv")
df = df.drop(columns=["Id"])
df = df[~df.SalesPrice.isnull()]
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave  None      Reg   
1          20       RL         80.0     9600   Pave  None      Reg   
2          60       RL         68.0    11250   Pave  None      IR1   
3          70       RL         60.0     9550   Pave  None      IR1   
4          60       RL         84.0    14260   Pave  None      IR1   

  LandContour LotConfig LandSlope    ...     PoolArea PoolQC Fence  \
0         Lvl    Inside       Gtl    ...            0   None  None   
1         Lvl       FR2       Gtl    ...            0   None  None   
2         Lvl    Inside       Gtl    ...            0   None  None   
3         Lvl    Corner       Gtl    ...            0   None  None   
4         Lvl       FR2       Gtl    ...            0   None  None   

  MiscFeature MiscVal  MoSold  YrSold  SaleType  SaleCondition SalesPrice  
0        None       0       2    2008        WD         Normal   208500.0  
1        None       0       5    2007        WD         Normal   181500.0  
2        None       0       9    2008        WD         Normal   223500.0  
3        None       0       2    2006        WD        Abnorml   140000.0  
4        None       0      12    2008        WD         Normal   250000.0  

[5 rows x 79 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 0 to 1459
Data columns (total 79 columns):
MSSubClass       1460 non-null int64
MSZoning         1460 non-null object
LotFrontage      1460 non-null float64
LotArea          1460 non-null int64
Street           1460 non-null object
Alley            1460 non-null object
LotShape         1460 non-null object
LandContour      1460 non-null object
LotConfig        1460 non-null object
LandSlope        1460 non-null object
Neighborhood     1460 non-null object
Condition1       1460 non-null object
Condition2       1460 non-null object
BldgType         1460 non-null object
HouseStyle       1460 non-null object
OverallQual      1460 non-null int64
OverallCond      1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
RoofStyle        1460 non-null object
RoofMatl         1460 non-null object
Exterior1st      1460 non-null object
Exterior2nd      1460 non-null object
MasVnrType       1460 no

# Exercise:
- Build a model to predict the SalesPrice using other columns as features. Split the dataset (1460 records) into training and test sets with 70/30 ratio. Take log of SalesPrice for prediction to reduce the impact of the outliers.
- What is the R2 score and rmse score? 
- Fit Lasso model. Test a few alpha values (>0) and see which gives a good r2 score. 

In [4]:
import numpy as np
from sklearn import preprocessing, model_selection, linear_model, metrics, pipeline

In [5]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 1)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.LinearRegression())
])


pipe.fit(X_train, y_train)

y_train_pred = pipe.predict(X_train)
y_test_pred = pipe.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))
est = pipe.steps[-1][-1]
pd.DataFrame({"feature": columns, "coefficient": est.coef_})

training r2 0.9494518297088844
test r2 -3.0715943247509893e+22
training rmse: 0.08606817747710854
test rmse: 76190488351.17424


feature   coefficient
0               MSSubClass  1.248159e-02
1              LotFrontage  1.963124e-03
2                  LotArea  3.151273e-02
3              OverallQual  5.675519e-02
4              OverallCond  3.468973e-02
5                YearBuilt  5.024337e-02
6             YearRemodAdd  1.540301e-02
7               MasVnrArea -3.995381e-03
8               BsmtFinSF1  4.530257e+10
9               BsmtFinSF2  1.612199e+10
10               BsmtUnfSF  4.288808e+10
11             TotalBsmtSF -4.382796e+10
12                1stFlrSF -1.112376e+11
13                2ndFlrSF -1.233520e+11
14            LowQualFinSF -1.362186e+10
15               GrLivArea  1.501695e+11
16            BsmtFullBath  8.699417e-03
17            BsmtHalfBath  1.619369e-03
18                FullBath  7.283211e-03
19                HalfBath  9.532332e-03
20            BedroomAbvGr  1.488495e-02
21            KitchenAbvGr -1.001364e-02
22            TotRmsAbvGrd -1.378536e-03
23              Fireplaces  1.032925e-02
24             GarageYrBlt -1.290452e-02
25              GarageCars  5.177021e-03
26              GarageArea  3.925258e-02
27              WoodDeckSF  1.009178e-02
28             OpenPorchSF  9.440899e-03
29           EnclosedPorch  1.197055e-02
..                     ...           ...
228          GarageCond_Gd  2.391434e-02
229        GarageCond_None  7.930320e+10
230          GarageCond_Po  3.245735e-02
231          GarageCond_TA  6.799388e-02
232           PavedDrive_P -3.139496e-03
233           PavedDrive_Y  2.371550e-03
234              PoolQC_Fa  1.114522e+03
235              PoolQC_Gd  1.083305e-02
236            PoolQC_None  3.260802e-02
237             Fence_GdWo -1.957178e-03
238            Fence_MnPrv  6.403923e-04
239             Fence_MnWw  2.841771e-03
240             Fence_None  5.779266e-04
241       MiscFeature_None  1.070099e-01
242       MiscFeature_Othr  2.083778e-02
243       MiscFeature_Shed  9.071565e-02
244       MiscFeature_TenC  3.388132e-21
245           SaleType_CWD  9.831190e-04
246           SaleType_Con  5.665779e-03
247         SaleType_ConLD  7.680893e-03
248         SaleType_ConLI -4.849315e-03
249         SaleType_ConLw -3.542900e-04
250           SaleType_New  4.478741e-02
251           SaleType_Oth  8.874655e-03
252            SaleType_WD -5.639076e-03
253  SaleCondition_AdjLand  1.772881e-03
254   SaleCondition_Alloca  1.122475e-02
255   SaleCondition_Family  9.747744e-04
256   SaleCondition_Normal  2.903473e-02
257  SaleCondition_Partial -1.682472e-02

[258 rows x 2 columns]

In [6]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 1)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Lasso(alpha=0.01, random_state=1))
])


pipe.fit(X_train, y_train)

y_train_pred = pipe.predict(X_train)
y_test_pred = pipe.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))
est = pipe.steps[-1][-1]


summary = pd.DataFrame({"feature": columns, "coefficient": est.coef_})
summary["coef_mag"] = np.abs(summary.coefficient)
summary = summary.sort_values("coef_mag", ascending = False)
summary = summary[summary.coef_mag>0]
summary

training r2 0.8940480673803641
test r2 0.8785389987905549
training rmse: 0.12460754170603186
test rmse: 0.15150859029885974


feature  coefficient  coef_mag
3             OverallQual     0.104135  0.104135
15              GrLivArea     0.095563  0.095563
25             GarageCars     0.055896  0.055896
5               YearBuilt     0.029465  0.029465
39            MSZoning_RM    -0.026432  0.026432
4             OverallCond     0.023309  0.023309
6            YearRemodAdd     0.021713  0.021713
70   Neighborhood_NridgHt     0.020148  0.020148
60   Neighborhood_Crawfor     0.019489  0.019489
16           BsmtFullBath     0.018030  0.018030
90        Condition2_PosN    -0.015700  0.015700
0              MSSubClass    -0.014961  0.014961
178      BsmtFinType1_Unf    -0.014265  0.014265
210      FireplaceQu_None    -0.013760  0.013760
157      Foundation_PConc     0.012625  0.012625
18               FullBath     0.011998  0.011998
2                 LotArea     0.011859  0.011859
76   Neighborhood_StoneBr     0.011295  0.011295
22           TotRmsAbvGrd     0.011208  0.011208
80        Condition1_Norm     0.010955  0.010955
27             WoodDeckSF     0.010850  0.010850
235             PoolQC_Gd    -0.010616  0.010616
169       BsmtExposure_Gd     0.010399  0.010399
202       Functional_Maj2    -0.009603  0.009603
213     GarageType_Attchd     0.009391  0.009391
194          CentralAir_Y     0.009121  0.009121
119   Exterior1st_BrkFace     0.009079  0.009079
163         BsmtQual_None    -0.008641  0.008641
23             Fireplaces     0.008159  0.008159
250          SaleType_New     0.007931  0.007931
172     BsmtExposure_None    -0.007832  0.007832
12               1stFlrSF     0.007722  0.007722
118   Exterior1st_BrkComm    -0.007699  0.007699
49      LotConfig_CulDSac     0.007518  0.007518
61   Neighborhood_Edwards    -0.006650  0.006650
31            ScreenPorch     0.005994  0.005994
171       BsmtExposure_No    -0.005965  0.005965
151          ExterQual_TA    -0.005217  0.005217
152          ExterCond_Fa    -0.004778  0.004778
11            TotalBsmtSF     0.004723  0.004723
96         BldgType_Twnhs    -0.004638  0.004638
28            OpenPorchSF     0.004604  0.004604
207        Functional_Typ     0.004278  0.004278
201        KitchenQual_TA    -0.004073  0.004073
193          HeatingQC_TA    -0.003568  0.003568
190          HeatingQC_Fa    -0.003450  0.003450
58   Neighborhood_ClearCr     0.002836  0.002836
206        Functional_Sev    -0.002447  0.002447
127    Exterior1st_Stucco    -0.002415  0.002415
21           KitchenAbvGr    -0.002275  0.002275
116      RoofMatl_WdShngl     0.001826  0.001826
174      BsmtFinType1_GLQ     0.001595  0.001595
45           LotShape_Reg    -0.001593  0.001593
142    Exterior2nd_Stucco    -0.001218  0.001218
105       RoofStyle_Gable    -0.001104  0.001104
69   Neighborhood_NoRidge     0.000791  0.000791
44           LotShape_IR3    -0.000444  0.000444
64   Neighborhood_MeadowV    -0.000369  0.000369
255  SaleCondition_Family    -0.000161  0.000161

In [7]:
summary = pd.DataFrame({"feature": columns, "coefficient": est.coef_})
summary["coef_mag"] = np.abs(summary.coefficient)
summary = summary.sort_values("coef_mag", ascending = False)
summary = summary[summary.coef_mag>0]
summary

feature  coefficient  coef_mag
3             OverallQual     0.104135  0.104135
15              GrLivArea     0.095563  0.095563
25             GarageCars     0.055896  0.055896
5               YearBuilt     0.029465  0.029465
39            MSZoning_RM    -0.026432  0.026432
4             OverallCond     0.023309  0.023309
6            YearRemodAdd     0.021713  0.021713
70   Neighborhood_NridgHt     0.020148  0.020148
60   Neighborhood_Crawfor     0.019489  0.019489
16           BsmtFullBath     0.018030  0.018030
90        Condition2_PosN    -0.015700  0.015700
0              MSSubClass    -0.014961  0.014961
178      BsmtFinType1_Unf    -0.014265  0.014265
210      FireplaceQu_None    -0.013760  0.013760
157      Foundation_PConc     0.012625  0.012625
18               FullBath     0.011998  0.011998
2                 LotArea     0.011859  0.011859
76   Neighborhood_StoneBr     0.011295  0.011295
22           TotRmsAbvGrd     0.011208  0.011208
80        Condition1_Norm     0.010955  0.010955
27             WoodDeckSF     0.010850  0.010850
235             PoolQC_Gd    -0.010616  0.010616
169       BsmtExposure_Gd     0.010399  0.010399
202       Functional_Maj2    -0.009603  0.009603
213     GarageType_Attchd     0.009391  0.009391
194          CentralAir_Y     0.009121  0.009121
119   Exterior1st_BrkFace     0.009079  0.009079
163         BsmtQual_None    -0.008641  0.008641
23             Fireplaces     0.008159  0.008159
250          SaleType_New     0.007931  0.007931
172     BsmtExposure_None    -0.007832  0.007832
12               1stFlrSF     0.007722  0.007722
118   Exterior1st_BrkComm    -0.007699  0.007699
49      LotConfig_CulDSac     0.007518  0.007518
61   Neighborhood_Edwards    -0.006650  0.006650
31            ScreenPorch     0.005994  0.005994
171       BsmtExposure_No    -0.005965  0.005965
151          ExterQual_TA    -0.005217  0.005217
152          ExterCond_Fa    -0.004778  0.004778
11            TotalBsmtSF     0.004723  0.004723
96         BldgType_Twnhs    -0.004638  0.004638
28            OpenPorchSF     0.004604  0.004604
207        Functional_Typ     0.004278  0.004278
201        KitchenQual_TA    -0.004073  0.004073
193          HeatingQC_TA    -0.003568  0.003568
190          HeatingQC_Fa    -0.003450  0.003450
58   Neighborhood_ClearCr     0.002836  0.002836
206        Functional_Sev    -0.002447  0.002447
127    Exterior1st_Stucco    -0.002415  0.002415
21           KitchenAbvGr    -0.002275  0.002275
116      RoofMatl_WdShngl     0.001826  0.001826
174      BsmtFinType1_GLQ     0.001595  0.001595
45           LotShape_Reg    -0.001593  0.001593
142    Exterior2nd_Stucco    -0.001218  0.001218
105       RoofStyle_Gable    -0.001104  0.001104
69   Neighborhood_NoRidge     0.000791  0.000791
44           LotShape_IR3    -0.000444  0.000444
64   Neighborhood_MeadowV    -0.000369  0.000369
255  SaleCondition_Family    -0.000161  0.000161

In [8]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 1)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Ridge(alpha=40, random_state=1))
])


pipe.fit(X_train, y_train)

y_train_pred = pipe.predict(X_train)
y_test_pred = pipe.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))
est = pipe.steps[-1][-1]


summary = pd.DataFrame({"feature": columns, "coefficient": est.coef_})
summary["coef_mag"] = np.abs(summary.coefficient)
summary = summary.sort_values("coef_mag", ascending = False)
summary = summary[summary.coef_mag>0]
summary

training r2 0.9388343768471614
test r2 0.8776469231977766
training rmse: 0.09467680401312785
test rmse: 0.15206395314993107


feature  coefficient  coef_mag
3              OverallQual     0.060435  0.060435
110       RoofMatl_CompShg     0.058395  0.058395
15               GrLivArea     0.053418  0.053418
32                PoolArea     0.040313  0.040313
114       RoofMatl_Tar&Grv     0.038677  0.038677
12                1stFlrSF     0.036038  0.036038
116       RoofMatl_WdShngl     0.034974  0.034974
4              OverallCond     0.031776  0.031776
13                2ndFlrSF     0.031732  0.031732
70    Neighborhood_NridgHt     0.029292  0.029292
236            PoolQC_None     0.028548  0.028548
25              GarageCars     0.028194  0.028194
90         Condition2_PosN    -0.027839  0.027839
5                YearBuilt     0.026332  0.026332
80         Condition1_Norm     0.025440  0.025440
235              PoolQC_Gd    -0.025355  0.025355
2                  LotArea     0.025239  0.025239
256   SaleCondition_Normal     0.025230  0.025230
60    Neighborhood_Crawfor     0.024447  0.024447
76    Neighborhood_StoneBr     0.023947  0.023947
48         LandContour_Lvl     0.022447  0.022447
61    Neighborhood_Edwards    -0.020933  0.020933
201         KitchenQual_TA    -0.020340  0.020340
64    Neighborhood_MeadowV    -0.018959  0.018959
111       RoofMatl_Membran     0.018696  0.018696
202        Functional_Maj2    -0.018465  0.018465
164            BsmtQual_TA    -0.018141  0.018141
18                FullBath     0.017977  0.017977
26              GarageArea     0.017562  0.017562
38             MSZoning_RL     0.017512  0.017512
..                     ...          ...       ...
245           SaleType_CWD     0.001277  0.001277
216     GarageType_CarPort    -0.001220  0.001220
239             Fence_MnWw    -0.001195  0.001195
123    Exterior1st_ImStucc     0.001178  0.001178
7               MasVnrArea    -0.001156  0.001156
103      HouseStyle_SFoyer    -0.001090  0.001090
232           PavedDrive_P    -0.001026  0.001026
93         Condition2_RRNn    -0.000975  0.000975
220       GarageFinish_RFn    -0.000786  0.000786
136    Exterior2nd_HdBoard    -0.000766  0.000766
133    Exterior2nd_BrkFace    -0.000701  0.000701
158        Foundation_Slab     0.000645  0.000645
132    Exterior2nd_Brk Cmn    -0.000627  0.000627
223          GarageQual_Gd    -0.000584  0.000584
17            BsmtHalfBath    -0.000583  0.000583
209         FireplaceQu_Gd     0.000510  0.000510
253  SaleCondition_AdjLand     0.000495  0.000495
131    Exterior2nd_AsphShn    -0.000438  0.000438
147        MasVnrType_None    -0.000418  0.000418
255   SaleCondition_Family     0.000384  0.000384
228          GarageCond_Gd     0.000356  0.000356
138    Exterior2nd_MetalSd    -0.000347  0.000347
134     Exterior2nd_CBlock     0.000199  0.000199
120     Exterior1st_CBlock     0.000199  0.000199
95         BldgType_Duplex     0.000190  0.000190
177       BsmtFinType1_Rec     0.000190  0.000190
85         Condition1_RRNe    -0.000187  0.000187
88         Condition2_Norm    -0.000089  0.000089
173       BsmtFinType1_BLQ    -0.000041  0.000041
249         SaleType_ConLw    -0.000026  0.000026

[252 rows x 3 columns]

In [9]:
from sklearn import feature_selection

In [18]:
scaler = preprocessing.StandardScaler()
X_train_std = scaler.fit_transform(X_train)
_, pvals = feature_selection.f_regression(X_train_std, y_train)
summary = pd.DataFrame({"feature": columns, "pval": pvals})
summary = summary.sort_values("pval")
summary = summary[summary.pval < 0.01] 
# pval < 0.01 are considered significant
# closer to 0 is more significant features
summary.iloc[:10,:]

/anaconda3/lib/python3.6/site-packages/sklearn/feature_selection/univariate_selection.py:299: RuntimeWarning: invalid value encountered in true_divide
  corr /= X_norms
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
/anaconda3/lib/python3.6/site-packages/scipy/stats/_distn_infrastructure.py:1821: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= self.a)


feature           pval
3     OverallQual  1.543610e-232
15      GrLivArea  1.502663e-147
25     GarageCars  8.980028e-140
26     GarageArea  9.673012e-121
18       FullBath  7.435435e-106
11    TotalBsmtSF   4.707153e-98
151  ExterQual_TA   1.996594e-96
12       1stFlrSF   1.957045e-95
5       YearBuilt   2.946855e-85
6    YearRemodAdd   3.840149e-80

In [21]:
target = "SalesPrice"
X = df.drop(columns=[target])
X = pd.get_dummies(X, drop_first=True)
columns = X.columns
X = X.values.astype("float")
y = np.log(df[target])

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y,
                           test_size = 0.3, random_state = 1)

pipe = pipeline.Pipeline([
    ("poly", preprocessing.PolynomialFeatures(degree=1
                                    , include_bias=False)),
    ("scaler", preprocessing.StandardScaler()),
    ("est", linear_model.Lasso(alpha=0.07, random_state=1))
])


pipe.fit(X_train, y_train)

y_train_pred = pipe.predict(X_train)
y_test_pred = pipe.predict(X_test)

print("training r2", metrics.r2_score(y_train, y_train_pred))
print("test r2", metrics.r2_score(y_test, y_test_pred))

print("training rmse:"
    , np.sqrt(metrics.mean_squared_error(y_train, y_train_pred)))
print("test rmse:"
      , np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))
est = pipe.steps[-1][-1]


summary = pd.DataFrame({"feature": columns, "coefficient": est.coef_})
summary["coef_mag"] = np.abs(summary.coefficient)
summary = summary.sort_values("coef_mag", ascending = False)
summary = summary[summary.coef_mag>0]
summary

training r2 0.7395929485859631
test r2 0.746971064732701
training rmse: 0.19535127802592958
test rmse: 0.2186774107558645


feature  coefficient  coef_mag
3    OverallQual     0.140629  0.140629
15     GrLivArea     0.069171  0.069171
25    GarageCars     0.055262  0.055262
11   TotalBsmtSF     0.014343  0.014343
5      YearBuilt     0.012730  0.012730
6   YearRemodAdd     0.009382  0.009382
12      1stFlrSF     0.006673  0.006673
39   MSZoning_RM    -0.005534  0.005534
23    Fireplaces     0.000774  0.000774